---
title: "Week 3 -- Visualizing tabular data with ggplot2"
---

> #### Learning objectives
>
> * Understand ggplot2's grammar of graphics
> * Apply the grammar to create a variety of plots including scatter plots, box plots and histograms
> * Learn about how R handles categorical data

# Visualizing data with ggplot2

In this session we will focus our attention on the popular **ggplot2** package
for visualizing tabular data by creating various kinds of plot.

## METABRIC data set

We'll be using an extended version of the METABRIC data set in which columns
have been added for the mRNA expression values for selected genes, including
estrogen receptor alpha (ESR1), progesterone receptor (PGR), GATA3 and FOXA1.

The METABRIC study characterized the genomic mutations and gene expression
profiles for over 2000 primary breast tumours. In addition to the gene
expression obtained using microarrays, genome-wide copy number profiles were
obtained using SNP microarrays and targeted sequencing was performed using a
panel of 40 driver-mutation genes.

[Curtis *et al.*, Nature 486:346-52, 2012](https://pubmed.ncbi.nlm.nih.gov/22522925)

[Pereira *et al.*, Nature Communications 7:11479, 2016](https://www.ncbi.nlm.nih.gov/pubmed/27161491)

Both the clinical data and the gene expression values were downloaded from
[cBioPortal](https://www.cbioportal.org/study/summary?id=brca_metabric).

These have been combined into a single table using techniques we will cover
later in the course. We've removed observations for patient tumour samples for
which expression data are not available.

In addition, we have replaced spaces in column names with underscores so we
don't have to constantly refer to these using backticks ( **`` ` ``** ). Note,
however, that even after renaming up some of the column names, we have one
column, '3-gene classifier', that is not strictly a valid name in R as it
starts with a number and contains a hyphen. Valid names for variables in R
contain only letters, numbers and the dot (**`.`**) and underscore (**`_`**)
characters and should not begin with a number. So when we refer to the
'3-gene classifier' column we'll have to use backticks, e.g.
``metabric$`3-gene_classifier` ``.

```{r message = FALSE}
library(tidyverse)
metabric <- read_csv("data/metabric_clinical_and_expression_data.csv")
metabric
```

# Our first ggplot - a scatter plot

First we'll create a simple scatter plot much like the ones we've created in
previous sessions using R's `plot()` function. We'll plot the expression of
estrogen receptor alpha (ESR1) against that of the transcription factor, GATA3.

```{r scatter_plot_1}
ggplot(data = metabric) +
  geom_point(mapping = aes(x = GATA3, y = ESR1))
```

We specified 3 things to create this plot:

1. The **data** -- needs to be a data frame (or a tibble)
2. The type of plot -- this is called a **geom** in ggplot2 terminology
3. The mapping of variables in the data to visual properties (called **aesthetics** in ggplot2) of objects in the plot

In this case, the type of plot is a **`geom_point`**, ggplot2's function for a
scatter plot, and the expression of the genes GATA3 and ESR1 (estrogen receptor
alpha) are mapped to the **`x`** and **`y`** coordinates, i.e. the positions of
points on the scatter plot.

Perhaps the x and y coordinates aren't quite what we imagine as having aesthetic
qualities. Other aesthetics that you might expect to be referred to in this way
include the size, shape and colour of the points.

# Aesthetic mappings

Look up the `geom_point` help page to see the list of aesthetics that are
available for this geom.

```{r eval = FALSE}
?geom_point
```

The help page lists all the aesthetics you can set for this particular type of
scatter plot. If you scroll down to the aesthetics section you'll see that the
**`x`** and **`y`** aesthetics are in bold face which tells us that these are
required, i.e. you have to supply them, otherwise you'll get an error message
from ggplot.

Other aesthetics such as size, shape, colour and transparency (alpha) are
optional.

So what about those other aesthetics? Colours are usually fun -- let's try to
use a colour aesthetic.

We need to set a mapping of a variable to colour just like we did for x and y.

```{r scatter_plot_2}
ggplot(data = metabric) +
  geom_point(mapping = aes(x = GATA3, y = ESR1, colour = ER_status))
```

What about colouring based on a continuous variable?

```{r scatter_plot_3}
ggplot(data = metabric) +
  geom_point(mapping = aes(x = GATA3, y = ESR1, colour = Nottingham_prognostic_index))
```

ggplot2 uses a colour scale for a continous variable but discrete colours for
discrete or categorical values.

Size is not usually a good aesthetic to map to a variable. This is particularly
the case where we have such a large number of observations.

```{r scatter_plot_4}
ggplot(data = metabric) +
  geom_point(mapping = aes(x = GATA3, y = ESR1, size = ER_status))
```

Some aesthetics, such as size, are designed to be used with continuous
variables. If you map these to a discrete variable as we have done here, then
ggplot warns you that this isn't such a good idea.

It would however be useful to be able to decrease the size of the points given
that there are around 2000 in this data set. We can do so by setting the size
to a single value rather than mapping it to one of the variables in the data
set - this has to be done outside the aesthetic mappings, i.e. outside the
`aes()` bit.

```{r scatter_plot_5}
ggplot(data = metabric) +
  geom_point(mapping = aes(x = GATA3, y = ESR1, colour = Nottingham_prognostic_index), size = 0.5)
```

Some aesthetics can only be used with categorical variables, e.g. shape.

```{r eval = FALSE}
ggplot(data = metabric) +
  geom_point(mapping = aes(x = GATA3, y = ESR1, shape = Survival_time))
```

```
Error: A continuous variable can not be mapped to shape
```

Here we set the shape based on the 3-gene classifier.

```{r scatter_plot_6}
ggplot(data = metabric) +
  geom_point(mapping = aes(x = GATA3, y = ESR1, shape = `3-gene_classifier`), size = 0.9)
```

Note that some of the patient samples have not been classified and ggplot has
removed those points with missing values for the 3-gene classifier.

It is usually preferable to use colours to distinguish between different
categories but sometimes these are used together where we want to show which
group a data point belongs to in two different categorical variables.

```{r scatter_plot_7}
ggplot(data = metabric) +
  geom_point(mapping = aes(x = GATA3, y = ESR1, colour = PAM50, shape = `3-gene_classifier`), size = 0.9)
```

Transparency can be useful when we have a large number of points as we can more
easily tell when points are overlaid, but like size, it is not usually mapped to
a variable and sits outside the `aes()`.

```{r scatter_plot_8}
ggplot(data = metabric) +
  geom_point(mapping = aes(x = GATA3, y = ESR1, colour = `3-gene_classifier`), size = 0.75, alpha = 0.5)
```

Colours are more commonly used than shapes, sizes or transparency.

Virtually every aspect of the plots we've created can be customized. We will
learn how to change the colours used as well as other display parameters later
in the course.

# ggplot2 grammar

The "gg" in ggplot2 stands for "grammar of graphics". This grammar is a coherent
system for describing and building graphs.

The basic template is given in the summary box below. This template can be used
to create any of the different types of graph possible with ggplot2.

```{block type = "rmdblock"}
**ggplot2 grammar of graphics**

**`ggplot(data = <DATA>) + <GEOM_FUNCTION>(mapping = aes(<MAPPINGS>))`**

`ggplot()` creates the plot object.

Layers can be added by adding geoms such as `geom_point()` using the `+`
operator that ggplot2 has overridden for this purpose.
```

# Another plot type - bar chart

Look up ggplot2 package help in RStudio to see what other geoms are available
(find the ggplot2 package in the Packages tab normally found in the bottom right
hand corner).

The METABRIC study redefined how we think about breast cancer by identifying and
characterizing several new subtypes, referred to as *integrative clusters*.
Let's create a bar chart of the number of patients whose cancers fall within
each subtype in the METABRIC cohort.

`geom_bar` is the geom we need and it requires a single aesthetic mapping of the
categorical variable of interest to `x`.

```{r bar_plot_1}
ggplot(data = metabric) +
  geom_bar(mapping = aes(x = Integrative_cluster))
```

The dark grey bars are a big ugly - what if we want each bar to be a different
colour?

```{r bar_plot_2}
ggplot(data = metabric) +
  geom_bar(mapping = aes(x = Integrative_cluster, colour = Integrative_cluster))
```

Colouring the edges wasn't quite what we had in mind. Look at the help for
`geom_bar` to see what other aesthetic we should have used.

```{r bar_plot_3}
ggplot(data = metabric) +
  geom_bar(mapping = aes(x = Integrative_cluster, fill = Integrative_cluster))
```

What happens if we fill with something other than the integrative cluster?

```{r bar_plot_4}
ggplot(data = metabric) +
  geom_bar(mapping = aes(x = Integrative_cluster, fill = ER_status))
```

We get a stacked bar plot.

Note the similarity in what we did here to what we did with the scatter plot
- there is a common grammar.

Let's try another stacked bar plot, this time with a categorical variable with
more than two categories.

```{r bar_plot_5}
ggplot(data = metabric) +
  geom_bar(mapping = aes(x = Integrative_cluster, fill = `3-gene_classifier`))
```

What if want all the bars to be the same colour but not dark grey, e.g. blue?

```{r bar_plot_6}
ggplot(data = metabric) +
  geom_bar(mapping = aes(x = Integrative_cluster, fill = "blue"))
```

That doesn't look right - why not?

You can set the aesthetics to a fixed value but this needs to be outside the
mapping, just like we did before for size and transparency in the scatter plots.

```{r bar_plot_7}
ggplot(data = metabric) +
  geom_bar(mapping = aes(x = Integrative_cluster), fill = "blue")
```

Setting this inside the `aes()` mapping told ggplot2 to map the colour aesthetic
to some variable in the data frame, one that doesn't really exist but which is
created on-the-fly with a value of "blue" for every observation (patient).

## Statistical transformations (ADVANCED)

You may have noticed that ggplot2 didn't just plot values from our data set but
had to do some calculation first for the bar chart, i.e. it had to sum the
number of observations in each category.

Each geom has a **statistical transformation**. In the case of the scatter plot,
`geom_point` uses the "identity" transformation which means just use the values
as they are, i.e. not really a transformation at all. The statistical
transformation for `geom_bar` is "count", which means it will count the number
of observations for each category in the variable mapped to the x aesthetic.

You can see which statistical transformation is being used by a geom by looking
at the `stat` argument in the help page for that geom.

There are some circumstances where you'd want to change the `stat`, for example
if we already had count values in our table.

```{r}
counts <- tibble(
  `3-gene_classifier` = c("ER-/HER2-", "ER+/HER2- High Prolif", "ER+/HER2- Low Prolif", "HER2+"),
  count = c(290, 603, 619, 188)
)
counts
```

```{r bar_plot_8}
ggplot(data = counts) +
  geom_bar(mapping = aes(x = `3-gene_classifier`, y = count), stat = "identity")
```

# Multiple layers

Consider again the ggplot2 grammar:

```
ggplot(data = <DATA>) +
  <GEOM_FUNCTION>(mapping = aes(<MAPPINGS>))
```

Why do we have the ggplot part and geom part joined by a "+" symbol?

The **'+'** operator has been overridden by ggplot2 to add layers and other
components (scales, themes, etc.) to the plot.

We can have multiple geoms in the same plot. Each is a different layer.

Let's create a plot with two geoms.

```{r scatter_line_plot_1}
ggplot(data = metabric) +
  geom_point(mapping = aes(x = GATA3, y = ESR1)) +
  geom_smooth(mapping = aes(x = GATA3, y = ESR1))
```

What is **`geom_smooth()`** doing? Look up the help page.

Note that the shaded area represents the standard error bounds on the fitted
model.

It is important when breaking these ggplot statments across multiple lines to
ensure that the `+` is at the end of lines so that R knows you haven't finished
constructing your plot.

There is some annoying duplication of code in the code above. We've repeated the
exact same aesthetic mapping for both geoms. We can avoid this by putting the
mappings in the `ggplot()` function instead.

```{r scatter_line_plot_2}
ggplot(data = metabric, mapping = aes(x = GATA3, y = ESR1)) +
  geom_point() +
  geom_smooth()
```

Let's make the plot look a bit prettier by reducing the size of the points and
making them transparent. We're not mapping `size` or `alpha` to any variables,
just setting them to constant values, and we only want these settings to apply
to the points, so we set them inside `geom_point()`.

```{r scatter_line_plot_3}
ggplot(data = metabric, mapping = aes(x = GATA3, y = ESR1)) +
  geom_point(size = 0.5, alpha = 0.5) +
  geom_smooth()
```

Aesthetics set in the `ggplot` function are treated as global mappings for the
plot and are inherited by all geoms added to the plot.

Let's say we still want to colour the points by ER status.

```{r scatter_line_plot_4}
ggplot(data = metabric, mapping = aes(x = GATA3, y = ESR1, colour = ER_status)) +
  geom_point(size = 0.5, alpha = 0.5) +
  geom_smooth()
```

Well, that wasn't expected, although it's pretty neat. We only wanted to apply
the colour aesthetic to the points.

We can add to the aesthetics in the geom function to apply them just to that
geom.

```{r scatter_line_plot_5}
ggplot(data = metabric, mapping = aes(x = GATA3, y = ESR1)) +
  geom_point(mapping = aes(colour = ER_status), size = 0.5, alpha = 0.5) +
  geom_smooth()
```

Or if you have got your scatter plot just right, and you'd rather not interfere
with it while trying out adding another layer, you can use set the `inherit.aes`
option to `FALSE`.

```{r scatter_line_plot_6}
ggplot(data = metabric, mapping = aes(x = GATA3, y = ESR1, colour = ER_status)) +
  geom_point(size = 0.5, alpha = 0.5) +
  geom_smooth(mapping = aes(x = GATA3, y = ESR1), inherit.aes = FALSE)
```

# Box plots

Box plots are a particular favourite of Wednesday lunchtime seminars.

```{r box_plot_1}
ggplot(data = metabric, mapping = aes(x = ER_status, y = GATA3)) +
  geom_boxplot()
```

See `geom_boxplot` help to explain how the box and whiskers are constructed.

How about adding another layer to display all the points?

```{r box_plot_2}
ggplot(data = metabric, mapping = aes(x = ER_status, y = GATA3)) +
  geom_boxplot() +
  geom_point()
```

The `geom_point` help points to `geom_jitter` as more suitable when one of the
variables is categorical.

```{r box_plot_3}
ggplot(data = metabric, mapping = aes(x = ER_status, y = GATA3)) +
  geom_boxplot() +
  geom_jitter()
```

Well, that's a bit of a mess. We need to reduce the spread or jitter and the
size and make the points transparent.

```{r box_plot_4}
ggplot(data = metabric, mapping = aes(x = ER_status, y = GATA3)) +
  geom_boxplot() +
  geom_jitter(width = 0.3, size = 0.5, alpha = 0.25)
```

Displaying points in this way makes much more sense when we only have a few
observations and where the box plot doesn't really portray that. Here it makes
less sense as we have very many observations.

Let's try a colour aesthetic to also look at how estrogen receptor expression
differs between progesterone receptor (PR) positive and negative tumours.

```{r box_plot_5}
ggplot(data = metabric, mapping = aes(x = ER_status, y = GATA3, colour = HER2_status)) +
  geom_boxplot()
```

# Time series data

The next type of plot we're going to look at is a **line plot**. Line plots are
a common visualization for time series data but we don't have data of this kind
in the METABRIC data set so we'll turn our attention to a more topical data set,
the reduction in different types of transport use over time in the UK in the
week leading up to and since the lockdown restrictions were introduced in
response to the coronavirus (COVID-19) pandemic.

These are data that the UK Government presents in its daily press conference and
published on its
[website](https://www.gov.uk/government/publications/slides-and-datasets-to-accompany-coronavirus-press-conference-24-april-2020).

```{r message = FALSE, warning = FALSE}
library(readxl)
transport_use <- read_excel("data/2020-04-24_COVID-19_Press_Conference_Data__1_.xlsx",
                            sheet = "Transport use",
                            col_names = c("date", "transport_type", "percentage"),
                            col_types = c("date", "text", "numeric"),
                            skip = 4,
                            n_max = 186)
transport_use$percentage <- transport_use$percentage * 100
transport_use
```

In this data set each row represents an observation of the current traffic
volume for various modes of transport (motor vehicle, bus, train, etc.) since
16 March as a percentage of the volume measured during the first week of
February.

So let's get a feel for this data set. Firstly, what modes of transport are
included and how many observations do we have for each?

```{r}
table(transport_use$transport_type)
```

We'll focus solely on rail use to start with.

```{r}
rail_use <- transport_use[transport_use$transport_type == "National rail", ]
rail_use
```

## Dates

There is something unfamiliar about the `date` column that needs some
explanation. Let's check the type and class of `date`.

```{r}
typeof(rail_use$date)
class(rail_use$date)
```

So the `date` column is stored as a double and is a suitable continuous variable
for a line plot but it's also got some special properties, afforded to by its
`POSIXct` class, that allows it to be treated as a date.

The `POSIXct` class is how R represents dates where POSIX is a standardization
scheme in the world of computers spanning multiple operating systems and
programming languages and the `ct` part stands for calendar time. We won't go
into the details here but note that the numeric representation of a date, i.e.
the `double` value, is the number of seconds since 1 January 1970. For example,
the date in our first row is:

```{r}
rail_use$date[1]
as.double(rail_use$date[1])
```

## Line plot

We're going to plot the volume of traffic on each date with each measurement
represented as a point on our graph and observations on successive dates
connected by a line. Looking at the geoms available in ggplot2 (see the list of
functions available for ggplot2 by navigating to the ggplot2 package in the
Packages tab in RStudio), it looks like **`geom_line`** is the one we need.

The aesthetics section in the help page for `geom_line` tells us we need an x
and y mapping. Our x aesthetic will be `date` and the y aesthetic will be
`percentage`.

```{r line_plot_1}
ggplot(data = rail_use) +
  geom_line(mapping = aes(x = date, y = percentage))
```

We can use another aesthetic, e.g. colour,  to plot a line for each of several
modes of transport.

```{r line_plot_2}
ggplot(data = transport_use) +
  geom_line(mapping = aes(x = date, y = percentage, colour = transport_type))
```

Another aesthetic available for `geom_line` is `linetype`.

```{r line_plot_3}
ggplot(data = transport_use) +
  geom_line(mapping = aes(x = date, y = percentage, linetype = transport_type))
```

# Histograms

```{r histogram_plot_1}
ggplot(data = metabric) +
  geom_histogram(mapping = aes(x = Age_at_diagnosis))
```

The warning message hints at picking a more optimal number of bins.

```{r histogram_plot_2}
ggplot(data = metabric) +
  geom_histogram(mapping = aes(x = Age_at_diagnosis), binwidth = 5)
```

Or can set the number of bins.

```{r histogram_plot_3}
ggplot(data = metabric) +
  geom_histogram(mapping = aes(x = Age_at_diagnosis), bins = 20)
```

These histograms are not aesthetically very pleasing - how about some better
aesthetics?

```{r histogram_plot_4}
ggplot(data = metabric) +
  geom_histogram(mapping = aes(x = Age_at_diagnosis), bins = 20, colour = "darkblue", fill = "grey")
```

# Categorical variables -- Factors

# The ggplot object -- a peek under the hood (ADVANCED)

Let's the build the plot up in stages.

```{r}
plot <- ggplot(data = metabric)
```

What is the plot object we've just created?

```{r}
typeof(plot)
```

```{r}
length(plot)
```

The plot object is a list containing 9 elements. It's actually a special type of
list.

```{r}
class(plot)
```

What are the 9 things in the list?

```{r}
names(plot)
```

So far we've constructed ggplots from data, layers (geoms) and aesthetic mappings.
We'll be looking at how to customize plots later, including changing the labels,
scales and theme.

The data element is just the metabric tibble that we provided to the `ggplot()`
function.

```{r}
plot$data
```

This `plot` list object is really only a specification for how ggplot2 should
create the plot. ggplot2 only renders the plot when we ask it to by printing it
out to the screen. We haven't added any layers yet so what does our plot look
like at this point?

```{r building_up_plot_1}
plot
```

```{r}
plot$mapping
plot$layers
```

Lets add an aesthetic mapping.

```{r building_up_plot_2}
plot <- ggplot(data = metabric, mapping = aes(x = Nottingham_prognostic_index, y = ESR1, colour = ER_status))
plot
```

ggplot2 has automatically added scales for x and y based on the ranges of
values for the Nottingham prognostic index and ESR1 expression. Still nothing
has been plotted as we haven't yet specified the type of plot (geom) to add as a
layer.

```{r}
plot$mapping
plot$layers
```

```{r building_up_plot_3}
plot <- plot + geom_point()
plot
```

```{r}
plot$layers
```

We touched on statistical transformations earlier. The `stat` associated with a
`geom_point` is `stat_identity` which leaves values unchanged. In the case of a
scatter plot, we already have the x and y values -- they don't need to be
transformed, just plotted on the x and y axes.

