---
title: "Week 3 -- Visualizing tabular data with ggplot2"
---

> #### Learning objectives
>
> * Understand ggplot2's grammar of graphics
> * Apply the grammar to create a variety of plots including scatter plots, box plots and histograms
> * Learn about how R handles categorical data

# Visualizing data with ggplot2

In this session we will focus our attention on the popular **ggplot2** package
for visualizing tabular data by creating various kinds of plot.

## METABRIC data set

We'll be using an extended version of the METABRIC data set in which columns
have been added for the mRNA expression values for selected genes, including
estrogen receptor alpha (ESR1), progesterone receptor (PGR), GATA3 and FOXA1.

The METABRIC study characterized the genomic mutations and gene expression
profiles for over 2000 primary breast tumours. In addition to the gene
expression obtained using microarrays, genome-wide copy number profiles were
obtained using SNP microarrays and targeted sequencing was performed using a
panel of 40 driver-mutation genes.

[Curtis *et al.*, Nature 486:346-52, 2012](https://pubmed.ncbi.nlm.nih.gov/22522925)

[Pereira *et al.*, Nature Communications 7:11479, 2016](https://www.ncbi.nlm.nih.gov/pubmed/27161491)

Both the clinical data and the gene expression values were downloaded from
[cBioPortal](https://www.cbioportal.org/study/summary?id=brca_metabric).

These have been combined into a single table using techniques we will cover
later in the course. We've removed observations for patient tumour samples for
which expression data are not available and have replaced spaces in column names
with underscores so we don't have to constantly refer to these using backticks
( **`** ).

```{r message = FALSE}
library(tidyverse)
metabric <- read_csv("data/metabric_clinical_and_expression_data.csv")
metabric
```

# Our first ggplot - a scatter plot

First we'll create a simple scatter plot much like the ones we've created in
previous sessions using R's `plot()` function. We'll plot the expression of
estrogen receptor alpha (ESR1) against that of the transcription factor, GATA3.

```{r scatter_plot_1}
ggplot(data = metabric) +
  geom_point(mapping = aes(x = GATA3, y = ESR1))
```

We specified 3 things to create this plot:

1. The **data** -- needs to be a data frame (or a tibble)
2. The type of plot -- this is called a **geom** in ggplot2 terminology
3. The mapping of variables in the data to visual properties (called **aesthetics** in ggplot2) of objects in the plot

In this case, the type of plot is a **`geom_point`**, ggplot2's function for a
scatter plot, and the expression of the genes GATA3 and ESR1 (estrogen receptor
alpha) are mapped to the **`x`** and **`y`** coordinates, i.e. the positions of
points on the scatter plot.

Perhaps the x and y coordinates aren't quite what we imagine as having aesthetic
qualities. Other aesthetics that you might expect to be referred to in this way
include the size, shape and colour of the points.

# Aesthetic mappings

Look up the `geom_point` help page to see the list of aesthetics that are
available for this geom.

```{r eval = FALSE}
?geom_point
```

The help page lists all the aesthetics you can set for this particular type of
scatter plot. If you scroll down to the aesthetics section you'll see that the
**`x`** and **`y`** aesthetics are in bold face which tells us that these are
required, i.e. you have to supply them, otherwise you'll get an error message
from ggplot.

Other aesthetics such as size, shape, colour and transparency (alpha) are
optional.

So what about those other aesthetics? Colours are usually fun -- let's try to
use a colour aesthetic.

We need to set a mapping of a variable to colour just like we did for x and y.

```{r scatter_plot_2}
ggplot(data = metabric) +
  geom_point(mapping = aes(x = GATA3, y = ESR1, colour = ER_status))
```

What about colouring based on a continuous variable?

```{r scatter_plot_3}
ggplot(data = metabric) +
  geom_point(mapping = aes(x = GATA3, y = ESR1, colour = Nottingham_prognostic_index))
```

ggplot2 uses a colour scale for a continous variable but discrete colours for
discrete or categorical values.

Size is not usually a good aesthetic to map to a variable. This is particularly
the case where we have such a large number of observations.

```{r scatter_plot_4}
ggplot(data = metabric) +
  geom_point(mapping = aes(x = GATA3, y = ESR1, size = ER_status))
```

Some aesthetics, such as size, are designed to be used with continuous
variables. If you map these to a discrete variable as we have done here, then
ggplot warns you that this isn't such a good idea.

It would however be useful to be able to decrease the size of the points given
that there are around 2000 in this data set. We can do so by setting the size
to a single value rather than mapping it to one of the variables in the data
set - this has to be done outside the aesthetic mappings, i.e. outside the
`aes()` bit.

```{r scatter_plot_5}
ggplot(data = metabric) +
  geom_point(mapping = aes(x = GATA3, y = ESR1, colour = Nottingham_prognostic_index), size = 0.5)
```

Some aesthetics can only be used with categorical variables, e.g. shape.

```{r eval = FALSE}
ggplot(data = metabric) +
  geom_point(mapping = aes(x = GATA3, y = ESR1, shape = Survival_time))
```

```
Error: A continuous variable can not be mapped to shape
```

Here we set the shape based on the 3-gene classifier.

```{r scatter_plot_6}
ggplot(data = metabric) +
  geom_point(mapping = aes(x = GATA3, y = ESR1, shape = `3_gene_classifier`), size = 0.9)
```

Note that some of the patient samples have not been classified and ggplot has
removed those points with missing values for the 3-gene classifier.

It is usually preferable to use colours to distinguish between different
categories but sometimes these are used together where we want to show which
group a data point belongs to in two different categorical variables.

```{r scatter_plot_7}
ggplot(data = metabric) +
  geom_point(mapping = aes(x = GATA3, y = ESR1, colour = PAM50, shape = `3_gene_classifier`), size = 0.9)
```

Transparency can be useful when we have a large number of points as we can more
easily tell when points are overlaid, but like size, it is not usually mapped to
a variable and sits outside the `aes()`.

```{r scatter_plot_8}
ggplot(data = metabric) +
  geom_point(mapping = aes(x = GATA3, y = ESR1, colour = `3_gene_classifier`), size = 0.75, alpha = 0.5)
```

Colours are more commonly used than shapes, sizes or transparency.

Virtually every aspect of the plots we've created can be customized. We will
learn how to change the colours used as well as other display parameters later
in the course.

# ggplot2 grammar

The "gg" in ggplot2 stands for "grammar of graphics". This grammar is a coherent
system for describing and building graphs.

The basic template is given in the summary box below. This template can be used
to create any of the different types of graph possible with ggplot2.

```{block type = "rmdblock"}
**ggplot2 grammar**

**`ggplot(data = <DATA>) + <GEOM_FUNCTION>(mapping = aes(<MAPPINGS>))`**

`ggplot()` creates the plot object.

Layers can be added by adding geoms such as `geom_point()` using the `+`
operator that ggplot2 has overridden for this purpose.
```

# Another plot type - bar chart

Look up ggplot2 package help in RStudio to see what other geoms are available
(find the ggplot2 package in the Packages tab normally found in the bottom right
hand corner).

The METABRIC study redefined how we think about breast cancer by identifying and
characterizing several new subtypes, referred to as *integrative clusters*.
Let's create a bar chart of the number of patients whose cancers fall within
each subtype in the METABRIC cohort.

`geom_bar` is the geom we need and it requires a single aesthetic mapping of the
categorical variable of interest to `x`.

```{r bar_plot_1}
ggplot(data = metabric) +
  geom_bar(mapping = aes(x = Integrative_cluster))
```

The dark grey bars are a big ugly - what if we want each bar to be a different
colour?

```{r bar_plot_2}
ggplot(data = metabric) +
  geom_bar(mapping = aes(x = Integrative_cluster, colour = Integrative_cluster))
```

Colouring the edges wasn't quite what we had in mind. Look at the help for
`geom_bar` to see what other aesthetic we should have used.

```{r bar_plot_3}
ggplot(data = metabric) +
  geom_bar(mapping = aes(x = Integrative_cluster, fill = Integrative_cluster))
```

What happens if we fill with something other than the integrative cluster?

```{r bar_plot_4}
ggplot(data = metabric) +
  geom_bar(mapping = aes(x = Integrative_cluster, fill = ER_status))
```

We get a stacked bar plot.

Note the similarity in what we did here to what we did with the scatter plot
- there is a common grammar.

What if want all the bars to be the same colour but not dark grey, e.g. blue?

```{r bar_plot_5}
ggplot(data = metabric) +
  geom_bar(mapping = aes(x = Integrative_cluster, fill = "blue"))
```

That doesn't look right - why not?

You can set the aesthetics to a fixed value but this needs to be outside the
mapping, just like we did before for size and transparency in the scatter plots.

```{r bar_plot_6}
ggplot(data = metabric) +
  geom_bar(mapping = aes(x = Integrative_cluster), fill = "blue")
```

Setting this inside the `aes()` mapping told ggplot2 to map the colour aesthetic
to some variable in the data frame, one that doesn't really exist but which is
created on-the-fly with a value of "blue" for every observation (patient).

# Multiple layers

Consider again the ggplot2 grammar:

```
ggplot(data = <DATA>) +
  <GEOM_FUNCTION>(mapping = aes(<MAPPINGS>))
```

Why do we have the ggplot part and geom part joined by a "+" symbol?

The **'+'** operator has been overridden by ggplot2 to add layers and other
components (scales, themes, etc.) to the plot.

We can have multiple geoms in the same plot. Each is a different layer.

Let's create a plot with two geoms.

```{r scatter_line_plot_1}
ggplot(data = metabric) +
  geom_point(mapping = aes(x = GATA3, y = ESR1)) +
  geom_smooth(mapping = aes(x = GATA3, y = ESR1))
```

What is **`geom_smooth()`** doing? Look up the help page.

Note that the shaded area represents the standard error bounds on the fitted
model.

It is important when breaking these ggplot statments across multiple lines to
ensure that the `+` is at the end of lines so that R knows you haven't finished
constructing your plot.

There is some annoying duplication of code in the code above. We've repeated the
exact same aesthetic mapping for both geoms. We can avoid this by putting the
mappings in the `ggplot()` function instead.

```{r scatter_line_plot_2}
ggplot(data = metabric, mapping = aes(x = GATA3, y = ESR1)) +
  geom_point() +
  geom_smooth()
```

Let's make the plot look a bit prettier by reducing the size of the points and
making them transparent. We're not mapping `size` or `alpha` to any variables,
just setting them to constant values, and we only want these settings to apply
to the points, so we set them inside `geom_point()`.

```{r scatter_line_plot_3}
ggplot(data = metabric, mapping = aes(x = GATA3, y = ESR1)) +
  geom_point(size = 0.5, alpha = 0.5) +
  geom_smooth()
```

Aesthetics set in the `ggplot` function are treated as global mappings for the
plot and are inherited by all geoms added to the plot.

Let's say we still want to colour the points by ER status.

```{r scatter_line_plot_4}
ggplot(data = metabric, mapping = aes(x = GATA3, y = ESR1, colour = ER_status)) +
  geom_point(size = 0.5, alpha = 0.5) +
  geom_smooth()
```

Well, that wasn't expected, although it's pretty neat. We only wanted to apply
the colour aesthetic to the points.

We can add to the aesthetics in the geom function to apply them just to that
geom.

```{r scatter_line_plot_5}
ggplot(data = metabric, mapping = aes(x = GATA3, y = ESR1)) +
  geom_point(mapping = aes(colour = ER_status), size = 0.5, alpha = 0.5) +
  geom_smooth()
```

Or if you have got your scatter plot just right, and you'd rather not interfere
with it while trying out adding another layer, you can use set the `inherit.aes`
option to `FALSE`.

```{r scatter_line_plot_6}
ggplot(data = metabric, mapping = aes(x = GATA3, y = ESR1, colour = ER_status)) +
  geom_point(size = 0.5, alpha = 0.5) +
  geom_smooth(mapping = aes(x = GATA3, y = ESR1), inherit.aes = FALSE)
```


