<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>Week 2 – Getting started with data in R</title>

<script src="site_libs/jquery-1.12.4/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/bootstrap.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<script src="site_libs/htmlwidgets-1.3/htmlwidgets.js"></script>
<link href="site_libs/datatables-css-0.0.0/datatables-crosstalk.css" rel="stylesheet" />
<script src="site_libs/datatables-binding-0.6/datatables.js"></script>
<link href="site_libs/dt-core-1.10.16/css/jquery.dataTables.min.css" rel="stylesheet" />
<link href="site_libs/dt-core-1.10.16/css/jquery.dataTables.extra.css" rel="stylesheet" />
<script src="site_libs/dt-core-1.10.16/js/jquery.dataTables.min.js"></script>
<link href="site_libs/crosstalk-1.0.0/css/crosstalk.css" rel="stylesheet" />
<script src="site_libs/crosstalk-1.0.0/js/crosstalk.min.js"></script>
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>

<link rel="stylesheet" href="styles.css" type="text/css" />



<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 51px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h2 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h3 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h4 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h5 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h6 {
  padding-top: 56px;
  margin-top: -56px;
}
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #ffffff;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  background: white;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row-fluid">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html"><img style="width: 120px; float: left; margin-top: 0px; margin-right: 20px;" src="images/CRUK_CI_logo.png"/>Introduction to R&nbsp;&nbsp;&nbsp;</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Contents
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="week0.html">Week 0 - Getting ready</a>
    </li>
    <li>
      <a href="week1.html">Week 1 - Introduction to R and RStudio</a>
    </li>
    <li>
      <a href="week2.html">Week 2 - Working with data in R</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/bioinformatics-core-shared-training/r-intro">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div class="fluid-row" id="header">



<h1 class="title toc-ignore">Week 2 – Getting started with data in R</h1>

</div>


<blockquote>
<h3 id="learning-objectives">Learning objectives</h3>
<ul>
<li>Create and run a script file containing your R code</li>
<li>More on vectors including extracting values or subsets and modifying their contents</li>
<li>Vector arithmetic</li>
<li>More sophisticated data structures but all built from vectors (lists and data frames)</li>
<li>Learn how R deals with missing values</li>
<li>Learn how to install and use packages that extend R’s basic functionality</li>
<li>Read data in tabular format from a spreadsheet</li>
<li>Calculate summary statistics on your tabular data</li>
<li>Introduce the <strong><code>tibble</code></strong>, arguably the most important data structure you will use in R</li>
</ul>
</blockquote>
<div id="introduction" class="section level1">
<h1>Introduction</h1>
<p>In this course, we’ll be mostly focusing on a set of R packages specifically designed for the most useful and common operations for interacting with and visualizing data, and particularly data in a tabular form. This collection of packages is known as the <strong>tidyverse</strong>.</p>
<p>Learning the <strong>tidyverse</strong> is more than just learning about some helpful functions to get certain tasks done. This collection of packages form a coherant system for data manipulation, exploration and visualization that share a common design philosophy. There is a certain elegance to code crafted using the tidyverse approach that is in stark contrast to the strange and often cryptic equivalent using traditional ‘base R’. This author only really became a true R convert on being introduced to the tidyverse.</p>
<p>Sadly, as much as we’d like, we can’t just cover the tidyverse alone and ignore the fundamentals of the R language. So this week we will look at some aspects of R that are crucial to understanding how R is handling your data, even though we will come back to some of these concepts in later weeks to show you how those same operations are more easily and elegantly carried out in the tidyverse.</p>
<p>We will also start to look at the most important data structure you’ll use with your data, assuming it is in tabular form, the <strong>data frame</strong>, and its superior tidyverse derivative, the <strong>tibble</strong>.</p>
</div>
<div id="scripts" class="section level1">
<h1>Scripts</h1>
<p>Up to now, we were mostly typing code in the Console pane at the <strong><code>&gt;</code></strong> prompt. This is a very interactive way of working with R but what if you want to save the commands you’ve typed for a future session working in R?</p>
<p>Instead we can create a script file containing our R commands that we can come back to later. This is the way most R coding is done so let’s have a go.</p>
<p>From the RStudio ‘<strong>File</strong>’ menu, select ‘<strong>New File</strong>’ and then ‘<strong>R Script</strong>’.</p>
<div class="figure">
<img src="images/RStudio_new_file_menu.png" style="width:50.0%" />

</div>
<p>You should now have a new file in the top left-hand side of RStudio for your new R script named ‘Untitled1’. The Console window no longer occupies the whole of left-hand side.</p>
<div class="figure">
<img src="images/RStudio_new_script.png" style="width:100.0%" />

</div>
<p>We can type code into this file just as we have done in the Console window at the command prompt.</p>
<p>Type in some of the commands from last week’s assignment. Do you notice that the file name on the tab for this pane is now highlighted in red and has an asterisk?</p>
<div class="figure">
<img src="images/RStudio_modified_script.png" style="width:50.0%" />

</div>
<p>This tells us that we haven’t yet saved our changes. There are various ways to do so just like in Word or Excel or other applications you’re familiar with, for example using the ‘<strong>Save</strong>’ option from the ‘<strong>File</strong>’ menu or clicking on the ‘<strong>Save</strong>’ button.</p>
<p>My preference by far is to use a keyboard shortcut. On a Mac this would be <kbd>cmd</kbd> + <kbd>S</kbd> (press the <kbd>cmd</kbd> key first and, while keeping this depressed, click the <kbd>S</kbd> key); on Windows you would do the same thing using <kbd>Ctrl</kbd> + <kbd>S</kbd>.</p>
<p>If the file already exists it will be saved without any further ado. As this is a new file, RStudio needs to know what you want to call it and in which folder on your computer you want it to be saved. You can choose these using the file dialog that appears. RStudio will add a ‘.R’ suffix if you don’t specify one.</p>
<p>It is a good idea to keep your scripts and assignment files for this course together in one folder or directory.</p>
<div id="running-scripts" class="section level2">
<h2>Running scripts</h2>
<p>Having typed an R command and hit the return key you’ll notice that the command isn’t actually run like it was in the console window. That’s because you’re writing your R code in an editor. To run a single line of code within your script you can press the ‘<strong>Run</strong>’ button at the top of the script.</p>
<div class="figure">
<img src="images/RStudio_run_script_command.png" />

</div>
<p>This will run the line of code on which the cursor is flashing or the next line of code if the cursor is on a blank or empty line.</p>
<p>The keyboard shortcut is more convenient in practice as you won’t have to stop typing at the keyboard to use your mouse. This is <kbd>cmd</kbd> + <kbd>return</kbd> on a Mac and <kbd>Ctrl</kbd> + <kbd>enter</kbd> on Windows.</p>
<p>Running a line in your script will automatically move the cursor onto the next command which can be very convenient as you’ll be able to run successive commands just by repeatedly clicking ‘<strong>Run</strong>’ or using the keyboard shortcut.</p>
<p>You can also run the entire script by clicking on the ‘<strong>Source</strong>’ button, a little to the right of the ‘<strong>Run</strong>’ button. More useful though is to run ‘<strong>Source with Echo</strong>’ as this will also display your commands and the outputs from these in the Console window.</p>
</div>
<div id="adding-comments-to-scripts" class="section level2">
<h2>Adding comments to scripts</h2>
<p>It is a very good idea to add comments to your code to explain what it’s doing or what you intended. This will help others to understand your code and more than likely even yourself when you come to revisit it a few weeks or months later.</p>
<p>Anything following a <strong><code>#</code></strong> symbol is a comment that R will ignore. Here’s an example of adding comments to our simple script.</p>
<div class="figure">
<img src="images/RStudio_script_comments.png" />

</div>
<p>Comments usually appear at the beginning of lines but can appear at the end of an R statement.</p>
<pre class="r"><code>days &lt;- c(1, 2, 4, 6, 8, 12, 16) # didn&#39;t manage to get a measurement on day 10</code></pre>
<p>It is also quite common when looking at R code to see lines of code commented out, usually replaced by another line that does something similar or makes a small change.</p>
<pre class="r"><code># random_numbers &lt;- rnorm(100, mean = 0, sd = 1)
random_numbers &lt;- rnorm(100, mean = 0, sd = 0.5)</code></pre>
</div>
</div>
<div id="vectors" class="section level1">
<h1>Vectors</h1>
<p>In week 1 we introduced <strong>vectors</strong>, the simplest type of data structure in R. An atomic vector is a collection of values or things of the same type in a given order. We created some last week using <strong><code>c()</code></strong> and the <strong><code>:</code></strong> operator.</p>
<pre class="r"><code>some_numbers &lt;- 1:10
days_of_the_week &lt;- c(&quot;Sunday&quot;, &quot;Monday&quot;, &quot;Tuesday&quot;, &quot;Wednesday&quot;, &quot;Thursday&quot;, &quot;Friday&quot;, &quot;Saturday&quot;)</code></pre>
<p>In the first example, an integer vector was created with 10 values from 1 to 10. <code>some_numbers</code> is a name that refers to the vector and can be used in other R statements while <code>1:10</code> is the vector object itself.</p>
<pre class="r"><code>squares &lt;- some_numbers ^ 2
squares</code></pre>
<pre><code>##  [1]   1   4   9  16  25  36  49  64  81 100</code></pre>
<p>A single value is known as a <strong>scalar</strong>. An example from last week was the number of samples in our experiment.</p>
<pre class="r"><code>samples &lt;- 8
samples</code></pre>
<pre><code>## [1] 8</code></pre>
<p>But as we saw last week, R doesn’t treat this single value any differently; it is still a vector, just one that has a length of 1.</p>
<p>Almost every object in R is a vector of one kind or another, or is constructed of vectors, so it’s really important to understand these well.</p>
<div id="combining-vectors" class="section level2">
<h2>Combining vectors</h2>
<p>The other way we’ve encountered for creating a vector is to use <strong><code>c()</code></strong>. This is actually a function and we can get help for it just as we can for any other function.</p>
<pre class="r"><code>?c</code></pre>
<p>From the help page you can see that <code>c</code> stands for ‘combine’ (or perhaps also ‘concatenate’, both terms are used in the documentation).</p>
<p>One of the most useful things about the help pages for functions are the examples that are given – you’ll need to scroll down to the bottom of the page to see them. These can be really helpful in demonstrating how a function works. You can very easily cut and paste these examples and run these in the console window as a way of experimenting with and learning about the function.</p>
<p>Let’s have a look at the first example. It’s slighly more complicated than what we did last week.</p>
<pre class="r"><code>c(1, 7:9)</code></pre>
<pre><code>## [1] 1 7 8 9</code></pre>
<p>This is actually combining two vectors, the first with a single value <code>1</code> and the second with values <code>7</code>, <code>8</code> and <code>9</code>. Here’s another example:</p>
<pre class="r"><code>cats &lt;- c(&quot;felix&quot;, &quot;kitty&quot;, &quot;tigger&quot;)
dogs &lt;- c(&quot;spot&quot;, &quot;snoopy&quot;)
cats_and_dogs &lt;- c(cats, dogs)
cats_and_dogs</code></pre>
<pre><code>## [1] &quot;felix&quot;  &quot;kitty&quot;  &quot;tigger&quot; &quot;spot&quot;   &quot;snoopy&quot;</code></pre>
</div>
<div id="coercion" class="section level2">
<h2>Coercion</h2>
<p>Atomic vectors must contain values that are all of the same type. A bit later on, we’ll introduce another type of data structure that doesn’t have this restriction – lists. First though, a look at back at one of the exercises from last week’s assignment in which we tried to create vectors of things that are of different types.</p>
<pre class="r"><code>integer_logical &lt;- c(1:5, c(TRUE, FALSE, TRUE))
integer_logical</code></pre>
<pre><code>## [1] 1 2 3 4 5 1 0 1</code></pre>
<pre class="r"><code>typeof(integer_logical)</code></pre>
<pre><code>## [1] &quot;integer&quot;</code></pre>
<p>Combining an integer vector, <code>1:5</code>, containing the numbers 1 through 5, with a logical vector results in an integer vector. The logical values have been <em>‘coerced’</em> into integers. But why the logical values and not the integers to produce a logical vector?</p>
<p>If you think about it, it makes more sense to convert logical values where <code>TRUE</code> and <code>FALSE</code> are usually represented in computers by the bits <code>1</code> and <code>0</code> respectively, i.e. <code>TRUE</code> and <code>FALSE</code> have natural and understandable equivalents in the world of integers. Which logical value would you give to the number 5 for example?</p>
<p>Similarly, integers get converted to doubles in this example:</p>
<pre class="r"><code>integer_double &lt;- c(3.4, 7, 2.3, 6:-3)
integer_double</code></pre>
<pre><code>##  [1]  3.4  7.0  2.3  6.0  5.0  4.0  3.0  2.0  1.0  0.0 -1.0 -2.0 -3.0</code></pre>
<pre class="r"><code>typeof(integer_double)</code></pre>
<pre><code>## [1] &quot;double&quot;</code></pre>
<p>Again, this makes more sense than converting doubles (numbers with a decimal point) to integers (whole numbers) and losing some of their precision.</p>
<p>Finally, it is really not obvious how to convert most character strings into either logical or numeric values, so when combining vectors that contain characters everything else gets ‘coerced’ into becoming characters also.</p>
<pre class="r"><code>we_all_want_to_be_characters &lt;- c(FALSE, 1:5, 23L, 3.23, 5.2e4, 1e-6, &quot;matt&quot;)
we_all_want_to_be_characters</code></pre>
<pre><code>##  [1] &quot;FALSE&quot; &quot;1&quot;     &quot;2&quot;     &quot;3&quot;     &quot;4&quot;     &quot;5&quot;     &quot;23&quot;    &quot;3.23&quot; 
##  [9] &quot;52000&quot; &quot;1e-06&quot; &quot;matt&quot;</code></pre>
<pre class="r"><code>typeof(we_all_want_to_be_characters)</code></pre>
<pre><code>## [1] &quot;character&quot;</code></pre>
</div>
<div id="extracting-subsets" class="section level2">
<h2>Extracting subsets</h2>
<p>One of the operations we do frequently on our data is to select subsets that are of particular interest. For example, we may be interested in the top 50 genes in a differential expression analysis for our RNA-seq experiment where those genes of interest are the ones with a log fold change above a certain value and with a p value below 0.01.</p>
<p>Having a good understanding of how to select a subset of values from a vector is going to be invaluable when we come to do the same for more complicated data structures so let’s take a look.</p>
<p>The main subsetting operator we’ll use is the square bracket, <strong><code>[]</code></strong>. Here’s an example.</p>
<pre class="r"><code>log2fc &lt;- c(2.3, -1, 0.48, 0.97, -0.02, 1.23)
log2fc[3]</code></pre>
<pre><code>## [1] 0.48</code></pre>
<p>We have a vector of five log<sub>2</sub> fold change values and we’ve chosen to select the third value.</p>
<p>If you’re familiar with other programming languages you will notice that the indexing scheme in R starts from 1, not 0. So the first element in the vector is referred to using the index 1, i.e. <code>log2fc[1]</code>.</p>
<p>Multiple values can be extracted by providing a vector of indices, e.g.</p>
<pre class="r"><code>log2fc[c(2, 4, 5)]</code></pre>
<pre><code>## [1] -1.00  0.97 -0.02</code></pre>
<p>You can also extract elements in a different order, e.g.</p>
<pre class="r"><code>log2fc[c(2, 5, 4)]</code></pre>
<pre><code>## [1] -1.00 -0.02  0.97</code></pre>
<p>It is possible to exclude values instead by providing negative indices.</p>
<pre class="r"><code>log2fc[c(-2, -3)]</code></pre>
<pre><code>## [1]  2.30  0.97 -0.02  1.23</code></pre>
<p>Which is usually more conveniently written using a vector of positive indices and negating these as follows.</p>
<pre class="r"><code>log2fc[-c(2, 3)]</code></pre>
<pre><code>## [1]  2.30  0.97 -0.02  1.23</code></pre>
<p>Finally, we can also subset our vector using a vector of logical values.</p>
<pre class="r"><code>log2fc[c(TRUE, FALSE, TRUE, TRUE, FALSE, FALSE)]</code></pre>
<pre><code>## [1] 2.30 0.48 0.97</code></pre>
<div id="conditional-subsetting" class="section level3">
<h3>Conditional subsetting</h3>
<p>You may be thinking that this last one seems very abstract and why on earth it could possibly be useful, but actually, it is probably the most frequently used way of selecting values of interest. To understand why, we’ll need to introduce the concept of logical operators.</p>
<p>Let’s say we’re interested in just the log<sub>2</sub> fold changes that are above a threshold of 0.5. We can test each of the values using the <strong><code>&gt;</code></strong> logical operator.</p>
<pre class="r"><code>log2fc &gt; 0.5</code></pre>
<pre><code>## [1]  TRUE FALSE FALSE  TRUE FALSE  TRUE</code></pre>
<p>This results in a logical vector containing <code>TRUE</code> and <code>FALSE</code> values for each element. The values at positions 1, 4 and 6 in our vector are above the threshold so result in <code>TRUE</code>, the others result in <code>FALSE</code>.</p>
<p>We can use this resulting vector to subset our original <code>log2fc</code> vector.</p>
<pre class="r"><code>log2fc_threshold &lt;- 0.5
above_threshold &lt;- log2fc &gt; log2fc_threshold
above_threshold</code></pre>
<pre><code>## [1]  TRUE FALSE FALSE  TRUE FALSE  TRUE</code></pre>
<pre class="r"><code>log2fc[above_threshold]</code></pre>
<pre><code>## [1] 2.30 0.97 1.23</code></pre>
<p>In practice, we wouldn’t really create a variable containing our logical vector signifying whether values are of interest. Instead we’d do this in a single step.</p>
<pre class="r"><code>log2fc[log2fc &gt; 0.5]</code></pre>
<pre><code>## [1] 2.30 0.97 1.23</code></pre>
<p>On the other hand, in a real R script, we might want to have a variable for the threshold that might be chosen by the user as an command-line argument and so could be different each time the script is run.</p>
<pre class="r"><code>log2fc_above_threshold &lt;- log2fc[log2fc &gt; log2fc_threshold]
log2fc_above_threshold</code></pre>
<pre><code>## [1] 2.30 0.97 1.23</code></pre>
<p>You can combine two or more conditions using <strong><code>&amp;</code></strong> if you want both conditions to be true or using <strong><code>|</code></strong> if either of the conditions is acceptable but not necessarily both.</p>
<pre class="r"><code># small fold changes
log2fc[log2fc &lt; 0.5 &amp; log2fc &gt; -0.5]</code></pre>
<pre><code>## [1]  0.48 -0.02</code></pre>
<pre class="r"><code># large fold changes
log2fc[log2fc &gt; 1 | log2fc &lt; -1]</code></pre>
<pre><code>## [1] 2.30 1.23</code></pre>
<p><code>&amp;</code> and <code>|</code> are the R versions of the AND and OR operations in Boolean algebra but applied to vectors.</p>
</div>
<div id="logical-operators" class="section level3">
<h3>Logical operators</h3>
<p>The following table lists the logical operators you can use in R.</p>
<table>
<thead>
<tr class="header">
<th>Operator</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>&lt;</td>
<td>less than</td>
</tr>
<tr class="even">
<td>&lt;=</td>
<td>less than or equal to</td>
</tr>
<tr class="odd">
<td>&gt;</td>
<td>greater than</td>
</tr>
<tr class="even">
<td>&gt;=</td>
<td>greater than or equal to</td>
</tr>
<tr class="odd">
<td>==</td>
<td>exactly equal to</td>
</tr>
<tr class="even">
<td>!=</td>
<td>not equal to</td>
</tr>
<tr class="odd">
<td>!x</td>
<td>NOT x</td>
</tr>
<tr class="even">
<td>x | y</td>
<td>x OR y</td>
</tr>
<tr class="odd">
<td>x &amp; y</td>
<td>x AND y</td>
</tr>
</tbody>
</table>
<p><code>x</code> and <code>y</code> in the last 3 of these operators are intended for logical values; if you apply them to other types, those other types will be coerced to logicals in exactly the same way we saw earlier.</p>
</div>
</div>
<div id="modifying-subsets" class="section level2">
<h2>Modifying subsets</h2>
<p>All subsetting operations can be combined with assignment. So we can modify or overwrite the values at specified position in our vector.</p>
<pre class="r"><code>some_numbers &lt;- 1:10
some_numbers[c(2, 4, 8)] &lt;- c(150, 34, -10)
some_numbers</code></pre>
<pre><code>##  [1]   1 150   3  34   5   6   7 -10   9  10</code></pre>
<p>And, as before, we could use a condition to decide which values to change. For example, you may decide that log<sub>2</sub> fold changes above 1 are somewhat unreliable with your detection method and so you’d like to put a cap on any values above this limit.</p>
<pre class="r"><code>log2fc[log2fc &gt; 1] &lt;- 1
log2fc</code></pre>
<pre><code>## [1]  1.00 -1.00  0.48  0.97 -0.02  1.00</code></pre>
</div>
</div>
<div id="vector-arithmetic" class="section level1">
<h1>Vector arithmetic</h1>
<p>Many operations in R are <strong>vectorized</strong>, which means that the operation is applied to an entire set of values at once. We’ve already seen lots of examples of this, like the following:</p>
<pre class="r"><code>some_numbers &lt;- 1:10
squares &lt;- some_numbers ^ 2
squares</code></pre>
<pre><code>##  [1]   1   4   9  16  25  36  49  64  81 100</code></pre>
<p>Here we used the <strong><code>^</code></strong> exponent operator to raise our numbers to the power of 2. This happened in a single operation, i.e. just one line of code. In other computer languages we might have had to write what is known as a loop in order to iterate over and perform the calculation for each value in turn.</p>
<p>Another way of writing the above statement to get the same result would be to multiply our numbers by themselves.</p>
<pre class="r"><code>squares &lt;- some_numbers * some_numbers
squares</code></pre>
<pre><code>##  [1]   1   4   9  16  25  36  49  64  81 100</code></pre>
<p>What actually happened here was slightly different though. Two vectors (actually the same vector in this case) of the same length were multiplied together. R did this <strong><em>element-by-element</em></strong>, which means that the first element from the first vector was multiplied by the first element of the second vector. Likewise, the second elements from each vector were multiplied by each other and so on.</p>
<p>Here’s another example involving two different vectors, but again these are of the same length.</p>
<pre class="r"><code>1:10 + 10:1</code></pre>
<pre><code>##  [1] 11 11 11 11 11 11 11 11 11 11</code></pre>
<p>Can you see why we get this result? (type <code>10:1</code> in the console if you’re not sure about what that gives).</p>
<p>Usually vector arithmetic involves two vectors of the same length or involves a vector and a scalar (a vector of length 1). One of the first examples from last week was the second scenario involving a vector and a single value.</p>
<pre class="r"><code>1:36 * 2</code></pre>
<pre><code>##  [1]  2  4  6  8 10 12 14 16 18 20 22 24 26 28 30 32 34 36 38 40 42 44 46
## [24] 48 50 52 54 56 58 60 62 64 66 68 70 72</code></pre>
<div id="vector-recycling-advanced" class="section level2">
<h2>Vector recycling (advanced)</h2>
<p>It is possible to perform calculations using two vectors of different sizes. When R runs out of values to use from the shorter of the two vectors, it wraps around to the beginning of that shorter one. For example, we can set alternate values in our set of numbers to be negative as follows:</p>
<pre class="r"><code>some_numbers * c(1, -1)</code></pre>
<pre><code>##  [1]   1  -2   3  -4   5  -6   7  -8   9 -10</code></pre>
<p>When R gets to the third element it has exhausted the shorter vector, <code>c(1, -1)</code>, so it goes back to the beginning, i.e. back to the first value, <code>1</code>. It uses the second vector five times in what is referred to as <strong><em>vector recycling</em></strong>.</p>
<p>You will probably never have to do something like this (why would you?) but without you knowing it you will carry out vector arithmetic using recycling very frequently. This is because a very common operation is to carry out a calculation on a vector using a single scalar value. For example, multiplying all values by a constant.</p>
<pre class="r"><code>heights_in_metres &lt;- c(1.86, 1.65, 1.72, 1.4, 1.79)
heights_in_centimetres &lt;- 100 * heights_in_metres
heights_in_centimetres</code></pre>
<pre><code>## [1] 186 165 172 140 179</code></pre>
<p>The multiplier of 100 is effectively being recycled and so is equivalent to us having written the following:</p>
<pre class="r"><code>heights_in_centimetres &lt;- c(100, 100, 100, 100, 100) * heights_in_metres</code></pre>
<p>Vector recycling in R requires that the length of the longer vector is an exact multiple of the length of the shorter. If not, R will give a warning.</p>
<pre class="r"><code>1:7 * c(1, -1)</code></pre>
<pre><code>## Warning in 1:7 * c(1, -1): longer object length is not a multiple of
## shorter object length</code></pre>
<pre><code>## [1]  1 -2  3 -4  5 -6  7</code></pre>
</div>
</div>
<div id="combining-data-of-different-types" class="section level1">
<h1>Combining data of different types</h1>
<p>Vectors are all very well and good but your data are almost certainly more complicated than an ordered set of values all of the same type. You’ve probably been working with Excel spreadsheets that contain some columns that are numerical while others contain names or character strings, e.g. the following table of Star Wars characters.</p>
<div id="htmlwidget-a9a6d1b485de3b04088d" style="width:100%;height:auto;" class="datatables html-widget"></div>
<script type="application/json" data-for="htmlwidget-a9a6d1b485de3b04088d">{"x":{"filter":"none","data":[["Luke Skywalker","C-3PO","R2-D2","Darth Vader","Leia Organa","Owen Lars","Beru Whitesun lars","R5-D4","Biggs Darklighter","Obi-Wan Kenobi","Anakin Skywalker","Wilhuff Tarkin","Chewbacca","Han Solo","Greedo","Jabba Desilijic Tiure","Wedge Antilles","Jek Tono Porkins","Yoda","Palpatine","Boba Fett","IG-88","Bossk","Lando Calrissian","Lobot","Ackbar","Mon Mothma","Arvel Crynyd","Wicket Systri Warrick","Nien Nunb","Qui-Gon Jinn","Nute Gunray","Finis Valorum","Jar Jar Binks","Roos Tarpals","Rugor Nass","Ric Olié","Watto","Sebulba","Quarsh Panaka","Shmi Skywalker","Darth Maul","Bib Fortuna","Ayla Secura","Dud Bolt","Gasgano","Ben Quadinaros","Mace Windu","Ki-Adi-Mundi","Kit Fisto","Eeth Koth","Adi Gallia","Saesee Tiin","Yarael Poof","Plo Koon","Mas Amedda","Gregar Typho","Cordé","Cliegg Lars","Poggle the Lesser","Luminara Unduli","Barriss Offee","Dormé","Dooku","Bail Prestor Organa","Jango Fett","Zam Wesell","Dexter Jettster","Lama Su","Taun We","Jocasta Nu","Ratts Tyerell","R4-P17","Wat Tambor","San Hill","Shaak Ti","Grievous","Tarfful","Raymus Antilles","Sly Moore","Tion Medon","Finn","Rey","Poe Dameron","BB8","Captain Phasma","Padmé Amidala"],[172,167,96,202,150,178,165,97,183,182,188,180,228,180,173,175,170,180,66,170,183,200,190,177,175,180,150,null,88,160,193,191,170,196,224,206,183,137,112,183,163,175,180,178,94,122,163,188,198,196,171,184,188,264,188,196,185,157,183,183,170,166,165,193,191,183,168,198,229,213,167,79,96,193,191,178,216,234,188,178,206,null,null,null,null,null,165],[77,75,32,136,49,120,75,32,84,77,84,null,112,80,74,1358,77,110,17,75,78.2,140,113,79,79,83,null,null,20,68,89,90,null,66,82,null,null,null,40,null,null,80,null,55,45,null,65,84,82,87,null,50,null,null,80,null,85,null,null,80,56.2,50,null,80,null,79,55,102,88,null,null,15,null,48,null,57,159,136,79,48,80,null,null,null,null,null,45],["male",null,null,"male","female","male","female",null,"male","male","male","male","male","male","male","hermaphrodite","male","male","male","male","male","none","male","male","male","male","female","male","male","male","male","male","male","male","male","male","male","male","male","male","female","male","male","female","male","male","male","male","male","male","male","female","male","male","male","male","male","female","male","male","female","female","female","male","male","male","female","male","male","female","female","male","female","male","male","female","male","male","male","female","male","male","female","male","none","female","female"],["Human","Droid","Droid","Human","Human","Human","Human","Droid","Human","Human","Human","Human","Wookiee","Human","Rodian","Hutt","Human","Human","Yoda's species","Human","Human","Droid","Trandoshan","Human","Human","Mon Calamari","Human","Human","Ewok","Sullustan","Human","Neimodian","Human","Gungan","Gungan","Gungan",null,"Toydarian","Dug",null,"Human","Zabrak","Twi'lek","Twi'lek","Vulptereen","Xexto","Toong","Human","Cerean","Nautolan","Zabrak","Tholothian","Iktotchi","Quermian","Kel Dor","Chagrian","Human","Human","Human","Geonosian","Mirialan","Mirialan","Human","Human","Human","Human","Clawdite","Besalisk","Kaminoan","Kaminoan","Human","Aleena",null,"Skakoan","Muun","Togruta","Kaleesh","Wookiee","Human",null,"Pau'an","Human","Human","Human","Droid",null,"Human"],["Tatooine","Tatooine","Naboo","Tatooine","Alderaan","Tatooine","Tatooine","Tatooine","Tatooine","Stewjon","Tatooine","Eriadu","Kashyyyk","Corellia","Rodia","Nal Hutta","Corellia","Bestine IV",null,"Naboo","Kamino",null,"Trandosha","Socorro","Bespin","Mon Cala","Chandrila",null,"Endor","Sullust",null,"Cato Neimoidia","Coruscant","Naboo","Naboo","Naboo","Naboo","Toydaria","Malastare","Naboo","Tatooine","Dathomir","Ryloth","Ryloth","Vulpter","Troiken","Tund","Haruun Kal","Cerea","Glee Anselm","Iridonia","Coruscant","Iktotch","Quermia","Dorin","Champala","Naboo","Naboo","Tatooine","Geonosis","Mirial","Mirial","Naboo","Serenno","Alderaan","Concord Dawn","Zolan","Ojom","Kamino","Kamino","Coruscant","Aleen Minor",null,"Skako","Muunilinst","Shili","Kalee","Kashyyyk","Alderaan","Umbara","Utapau",null,null,null,null,null,"Naboo"]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th>name<\/th>\n      <th>height<\/th>\n      <th>mass<\/th>\n      <th>gender<\/th>\n      <th>species<\/th>\n      <th>homeworld<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"columnDefs":[{"className":"dt-right","targets":[1,2]}],"order":[],"autoWidth":false,"orderClasses":false}},"evals":[],"jsHooks":[]}</script>
<p>Note that the first column contains character type data, the second and third columns contain numerical data (of type double) and the remaining columns contain the special type we briefly touched on last week, factors, that look like character types but have a limited set of values or categories.</p>
<div id="lists" class="section level2">
<h2>Lists</h2>
<p>R’s simplest structure that combines data of different types is a <strong>list</strong>. A list is a collection of vectors. It is also a vector itself but is a step up in complexity from the atomic vectors we’ve been looking at up till now. Vectors in a list can be of different types and different lengths.</p>
<pre class="r"><code>my_first_list &lt;- list(1:10, c(&quot;a&quot;, &quot;b&quot;, &quot;c&quot;), c(TRUE, FALSE), 100, c(1.3, 2.2, 0.75, 3.8))
my_first_list</code></pre>
<pre><code>## [[1]]
##  [1]  1  2  3  4  5  6  7  8  9 10
## 
## [[2]]
## [1] &quot;a&quot; &quot;b&quot; &quot;c&quot;
## 
## [[3]]
## [1]  TRUE FALSE
## 
## [[4]]
## [1] 100
## 
## [[5]]
## [1] 1.30 2.20 0.75 3.80</code></pre>
<p><code>my_first_list</code> has five elements and when printed out like this looks quite strange at first sight. Note how each of the elements of a list is referred to by an index within 2 sets of square brackets. This gives a clue to how you can access individual elements in the list.</p>
<pre class="r"><code>my_first_list[[2]]</code></pre>
<pre><code>## [1] &quot;a&quot; &quot;b&quot; &quot;c&quot;</code></pre>
<p>Elements in lists are normally named, e.g.</p>
<pre class="r"><code>genomics_instruments &lt;- list(sequencers = c(&quot;NovaSeq 6000&quot;, &quot;HiSeq 4000&quot;, &quot;NextSeq 500&quot;, &quot;MiSeq&quot;), liquid_handling_robots = c(&quot;Mosquito HV&quot;, &quot;Bravo&quot;))
genomics_instruments</code></pre>
<pre><code>## $sequencers
## [1] &quot;NovaSeq 6000&quot; &quot;HiSeq 4000&quot;   &quot;NextSeq 500&quot;  &quot;MiSeq&quot;       
## 
## $liquid_handling_robots
## [1] &quot;Mosquito HV&quot; &quot;Bravo&quot;</code></pre>
<p>We can still access the elements using the double square brackets but now we can use either the index (position) or the name.</p>
<pre class="r"><code>genomics_instruments[[1]]</code></pre>
<pre><code>## [1] &quot;NovaSeq 6000&quot; &quot;HiSeq 4000&quot;   &quot;NextSeq 500&quot;  &quot;MiSeq&quot;</code></pre>
<pre class="r"><code>genomics_instruments[[&quot;sequencers&quot;]]</code></pre>
<pre><code>## [1] &quot;NovaSeq 6000&quot; &quot;HiSeq 4000&quot;   &quot;NextSeq 500&quot;  &quot;MiSeq&quot;</code></pre>
<p>Even more conveniently we can use the <code>$</code> operator.</p>
<pre class="r"><code>genomics_instruments$sequencers</code></pre>
<pre><code>## [1] &quot;NovaSeq 6000&quot; &quot;HiSeq 4000&quot;   &quot;NextSeq 500&quot;  &quot;MiSeq&quot;</code></pre>
</div>
<div id="data-frames" class="section level2">
<h2>Data frames</h2>
<p>A much more useful data structure and the one we will mostly be using for the rest of the course is the <strong>data frame</strong>. This is actually a special type of list in which all the elements are vectors of the same length. The data frame is how R represents tabular data like the Star Wars table.</p>
<p>TODO</p>
<ul>
<li>need to explain that the starwars table is a data frame in the dplyr package in the tidyverse so need to load the tidyverse, library(tidyverse)</li>
<li>access columns in the data frame using $ notation just as we did for lists</li>
<li>selecting multiple columns using vectors or indices or names, e.g. starwars[c(1, 2, 3, 8, 10, 9)] or starwars[c(“name”, “height”, “mass”, “gender”, “species”, “homeworld”)</li>
<li>colnames(), dim(), nrow() and ncol() functions</li>
<li>access individual elements in matrix-like notation, e.g. [2, 6]</li>
<li>access subsets of rows, e.g. starwars[1:5,] or rows and columns, e.g. starwars[1:5, c(“name”, “species”)]</li>
<li>compute some summary statistics, e.g. summary(starwars[c(“name”, “height”, “mass”, “gender”, “species”, “homeworld”)]) or create a starwars_selected_columns object first to make this it bit easier to read</li>
</ul>
</div>
<div id="tibbles" class="section level2">
<h2>Tibbles</h2>
</div>
</div>
<div id="reading-data-into-r" class="section level1">
<h1>Reading data into R</h1>
<p>TODO</p>
<ul>
<li>details about the working directory from which files are read or are written to, getwd() and how to change this using setwd() or navigating to the directory using RStudio</li>
</ul>
</div>
<div id="missing-values-na" class="section level1">
<h1>Missing values (NA)</h1>
<p>TODO</p>
<ul>
<li>e.g. gender column in starwars table</li>
</ul>
</div>
<div id="summary" class="section level1">
<h1>Summary</h1>
</div>



</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open')
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_').toLowerCase();
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
